[Speculation] Automatic Finder for Speculative Units Placement

asleix opened the pull request:
### Description
A new feature for Speculation is added. Finder methods are implemented to allow for automatic placement of the speculative units, without the need of manually specifying al positions in the JSON file.

The pass can now be called with `--handshake-speculation="json-path=myjson.json automatic=true"` to automatically find the placements. Notice that `automatic=true` by default, so to manually specify the units, `automatic=false` is needed.

The feature is implemented in the class `PlacementFinder`, whose constructor and the method `findPlacements` are exposed. Many private methods are implemented to find the positions of all units, including saves, commits and save-commits. 

New tests are added to support this feature.

As this PR is a bit large, some discussions will be needed.

#### Control-flow-graph calculations
A helper struct to hold a BB's predecessor arcs is created:
```
struct BlockPredecessor {
  unsigned srcBB;
  unsigned dstBB;
  std::vector<OpPlacement> arcs;
};
```
Its purpose is to be able to map a BB with the arcs that lead to its predecessors using a `std::map<unsigned, std::vector<BlockPredecessor>>`. The function `getBlockPredecessors` is implemented to achieve this purpose. 

If this feature can be useful for `CFG.h`, it can be refactored in the future.

murphe67 commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,359 @@
+      if (placements.containsSave(res, succOp)) {
+        // A Commit is needed in front of Save Operations. This will be later
+        // converted into a SaveCommit for multiple loop speculation.
+        placements.addCommit(res, succOp);
```
I know this is what is described in the paper but can't we just remove the save placement now and add a save commit placement?

asleix replied:
Done.

murphe67 commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,359 @@
+  markedPaths.clear();
+  markSpeculativePaths(specPos.dstOp, placements, markedPaths);
+
+  // Remove commits that cannot be reached
```
Haoran has a lot of discussion of doing this properly in the thesis, but this approach seems pretty naive: is it the same as what he has written?

asleix replied:
For the record, Emmet and I discussed this in a meeting and decided to address optimizations in a future pass.

murphe67 commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,359 @@
+        continue;
+      } else if (isa<handshake::ConditionalBranchOp>(succOp)) {
+        // A SaveCommit is needed in front of the branch
+        placements.addSaveCommit(res, succOp);
```
this seems naive also: is this how the previous approach did it?
We definitely need some kind of save-commit placement minimization eventually

asleix replied:
For the record, Emmet and I discussed this in a meeting and decided to address optimizations in a future pass.

murphe67 commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,359 @@
+
+// Data structure to hold all arcs leading to a single BB predecessor
+// Note: srcBB and dstBB can be equal when the arcs are Backedges
+struct BlockPredecessor {
```
naming here confuses me "a single BB predecessor", then both src and dst are stored. 
what does "leading to a predecessor" mean? is it from a predecessor? if it has both a src and a dest does it hold "all" the arcs from the src or just the ones to the dest?
This is a group of arcs that have the same source and dest? I don't see how predecessors are involved

murphe67 replied:
I think this struct should be called a bbArc, and the vector is instead edges.

asleix replied:
The naming you suggest is very appropriate. The whole section has been refactored with your suggestions.

murphe67 commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,359 @@
+  funcOp->walk([&](Operation *op) {
+    for (Value operand : op->getOperands()) {
+      BBEndpoints endpoints;
+      if (isBackedge(operand, op, &endpoints) or
```
why are the endpoints being set by isBackedge? Is this normal? it looks very strange

asleix replied:
The function isBackedge is overloaded. If a reference to a BBEndpoints instance is passed, the endpoints of the edge (operand, op) is set. This allows for the later comparison in the same line `endpoints.srcBB != endpoints.dstBB`. This is according to the design of the `CFG.h` API.

murphe67 commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,359 @@
+    predecessor.srcBB = endpoints.srcBB;
+    predecessor.dstBB = endpoints.dstBB;
+    predecessor.arcs = arcs;
+    predecessors[endpoints.dstBB].push_back(predecessor);
```
so "predecessor" is a dict that maps a BB to the arc selected by the cmerge. I don't think this works as a name, and the arcs themselves definitely shouldn't be called predecessors.

asleix replied:
This is now called arc.

murphe67 commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,359 @@
+    unsigned countSpecInputs = 0;
+    for (BlockPredecessor pred : blockPredecessors) {
+      if (llvm::any_of(pred.arcs,
+                       [&](OpPlacement p) { return markedPaths.count(p); }))
```
this took me a long time to read since I don't write LLVM code, if you could comment it/abstract it better

asleix replied:
A comment has been added.

murphe67 commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,359 @@
+
+  // Whenever a BB has two speculative inputs, commit units are needed to avoid
+  // tokens going out-of-order. First, the block predecessor arcs are found
+  BBPredecessorMap predecessors = getBlockPredecessors(funcOp);
```
this also should not be called predecessors. I had no idea what you were iterating over, but now I see you're just iterating over all the BBs. I normally name these things something like bbToArcsMap

asleix replied:
The naming of this whole section has been improved to better describe what the objects do.

murphe67 commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,359 @@
+    BlockPredecessor predecessor;
+    predecessor.srcBB = endpoints.srcBB;
+    predecessor.dstBB = endpoints.dstBB;
+    predecessor.arcs = arcs;
```
I feel like the struct is the arc, and your current "arcs" variable are the edges that make up the arcs: there's only one arc between two BBs no matter how many data tokens, no?

asleix replied:
The struct is now called `BBArc` and the vector is called `edges`.

murphe67 commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,359 @@
+    if (countSpecInputs > 1) {
+      // Potential ordering issue, add commits
+      for (BlockPredecessor pred : blockPredecessors) {
+        for (OpPlacement arc : pred.arcs) {
```
just because a bb has 2 spec arcs doesn't mean every arc needs a commit, and in fact will produce incorrect functioning. Think adding an if statement on if "arc" is in markedPaths would work?
Also don't think the individual edges are arcs.

murphe67 replied:
Thinking about this more I think it actually functions fine, its just a resource waste.

asleix replied:
This has been addressed by checking if the input edge is speculative or not before adding a commit.

murphe67 reviewed (commented):
Hey Aleix, thanks for all the work on this!

I have some concerns with your datastructure naming that we can dig into, but I think apart from one small thing all of the logic and control flow looks great.

The other comment I would make is if you could comment all of the functions and classes with where they're called from and what they do, its not clear without a lot of reading how the control flow works between the functions.

asleix commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,359 @@
+  for (Value res : currOp->getResults()) {
+    for (Operation *succOp : res.getUsers()) {
+      OpPlacement arc = {res, succOp};
+      if (not markedPaths.count(arc)) {
```
Renamed to `PlacementSet`.

asleix commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,359 @@
+
+// DFS traversal to mark all operations that lead to Commit units
+static void markSpeculativePaths(Operation *currOp,
+                                 SpeculationPlacements &placements,
```
Renamed to `markSpeculativePathsForSaves` and `markSpeculativePathsForCommits`.

murphe67 commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
on line 322, can we remove mergeSaveCommits?

asleix replied:
Done.

murphe67 reviewed (commented):
Thanks Aleix, the new datastructure names look great! Some minor changes that are similiar to last time, and also some thoughts on the OpPlacement struct.

asleix commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -232,10 +247,10 @@ void PlacementFinder::findCommitsBetweenBBs() {
   // Remove commits that cannot be reached
-  PlacementList toRemove;
+  PlacementSet toRemove;
   for (OpPlacement arc : placements.getPlacements<handshake::SpecCommitOp>()) {
```
OpPlacement has been renamed to CFGEdge **only within this section of the code**.

asleix commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -145,83 +151,92 @@ using BBEndpointsMap =
-  PlacementList markedPaths;
-  markSpeculativePaths(specPos.dstOp, placements, markedPaths);
+  BBtoPredecessorArcsMap predecessors = getPredecessorArcs(funcOp);
+  PlacementSet markedPaths;
```
Done.

asleix commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -145,83 +151,92 @@ using BBEndpointsMap =
-  BBPredecessorMap predecessors = getBlockPredecessors(funcOp);
-  PlacementList markedPaths;
-  markSpeculativePaths(specPos.dstOp, placements, markedPaths);
+  BBtoPredecessorArcsMap predecessors = getPredecessorArcs(funcOp);
```
Agreed.

asleix commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -145,83 +151,92 @@ using BBEndpointsMap =
       if (not markedPaths.count(arc)) {
         markedPaths.insert(arc);
         // Stop traversal if a commit is reached
         if (not placements.containsCommit(res, succOp))
```
Agreed.

asleix commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -145,83 +151,92 @@ using BBEndpointsMap =
       BBEndpoints endpoints;
       if (isBackedge(operand, op, &endpoints) or
           endpoints.srcBB != endpoints.dstBB) {
         OpPlacement arc = {operand, op};
```
Renamed and added some comments for clarity.

asleix commented:
As per @murphe67 suggestions, I renamed and changed some of `SpeculationPlacements` methods to work better with edges.
- Added overloaded methods to `SpeculationPlacements` such that it works both with the two arguments `srcOpResult`, `*dstOp` and the struct `{srcOpResult, *dstOp}`.
- Renamed `OpPlacement` to `CFGEdge` in the section where the edges are used. However, I think that in other sections (such as in SpeculationPlacement.cpp) the name `OpPlacement` is still preferred. To achieve this, the line `using CFGEdge = OpPlacement;` is used, which achieves the best of both worlds.
- Improved overall comments when using these edges.

lucas-rami commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,389 @@
+  // There already exist save-commits which have been placed instead of
+  // consecutive save and commit units. Here, additional save commits are found
+  OpPlacement specPos = placements.getSpeculatorPlacement();
+  auto funcOp = specPos.dstOp->getParentOfType<handshake::FuncOp>();
```
Here is one of those cases where I think `auto` is fine because the return type is also on the right-hand-side of the assignment (as a more minor reason, adding the explicit type here probably causes a line break).

lucas-rami commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,389 @@
+  // If a save-commit is placed, then for correctness, every path from entry
+  // points to exit points in the Speculator BB should cross either the
+  // speculator or a save-commit
+  if (not placements.getPlacements<handshake::SpecSaveCommitOp>().empty()) {
```
use `!

lucas-rami commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,389 @@
+      if (not foundControlMerge) {
+        foundControlMerge = true;
+      } else {
+        controlMergeOp->emitError("Found many control merges in the same BB");
```
You can just do 
```cpp
return controlMergeOp->emitError("Found many control merges in the same BB");
```
This makes brackets for the if/else unnecessary.

lucas-rami reviewed (commented):
Thanks for the good work! I only left stylistic comments and have only one "design" comment, which I include below.

I guess I could have realized it in your previous PR, but isn't the `OpPlacement` data-structure just equivalent to a specific use of an SSA value? In our jargon, and equivalently, a single consumer of a token (leaving the possibility that the IR is not materialized and that a token may still have many consumers)?

If so, then it can simply be an `mlir::OpOperand&` (or pointer, if you need it as a key type in a map or set) instead of a pair of the SSA value and the consumer. These can both be retrieved on demand from the operand using the `.getValue()` and `.getOwner()` methods, respectively. Using the operand directly makes the code more readable in my opinion as it is a standard and widely-used MLIR type. It also removes the possibility of creating a wrong SSA value/consumer pair, as MLIR manages the data-type itself. I really think it is worth making the (mechanical) change throughout your code if I am correct. From previous experiences (i.e., me doing the same kind of thing in other parts of the codebase before realizing that the `mlir::OpOperand` was a thing), doing so can simplify the code significantly.

As for your question on whether part of this could be extracted into general-purpose CFG-analysis stuff that would go in our CFG support, I'm all for it if you both (@asleix and @murphe67) think it makes sense. Happy to discuss prior to the integration if necessary. I think it makes sense for it to go directly in non-experimental work (you're doing the "experimental" version right now in a way), so feel free to make a PR directly touching `CFG.h/cpp` if you decide to push for this.

murphe67 commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,389 @@
+}
+
+// Define a Control-Flow Graph Edge as the pair (srcOpResult, *dstOp)
+using CFGEdge = OpPlacement;
```
Personal opinion is if we drop the using, and switch to OpOperand but keep the variable names as "edge", we get both clarity on implementation type and clarity on how it maps to a pseudocode version

asleix replied:
Now using "using CFGEdge = OpOperand;". Once we move this part of the code to CFG.h, we can re-think it.

lucas-rami replied:
Sounds great, thanks!

asleix commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,389 @@
+  }
+}
+
+LogicalResult PlacementFinder::findCommitPositions() {
```
Added some conditions to verify correctness and now it can fail.

asleix commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,389 @@
+    return;
+
+  OpPlacement specPos = placements.getSpeculatorPlacement();
+  auto specBB = getLogicBB(specPos.dstOp);
```
Totally agree!

asleix commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -0,0 +1,389 @@
+
+  for (Value res : currOp->getResults()) {
+    for (Operation *succOp : res.getUsers()) {
+      if (auto succOpBB = getLogicBB(succOp); !succOpBB || succOpBB != specBB) {
```
Thanks for the nice suggestion! I wasn't very sure about this, your idea is a good improvement.

asleix commented:
Thanks @lucas-rami for the nice comments! I have addressed all of them.

The major structural change is the removal of `OpPlacement` and the use of `mlir::OpOperand` instead. The sets and maps now use `OpOperand *`, and the behavior is unchanged.

Some of the syntax may have become a bit more verbose. To traverse the IR in a DFS manner, now I use 
```
for (Operation *succOp : currOp->getUsers()) {
    for (OpOperand &dstOpOperand : succOp->getOpOperands()) {
      // Iterate the operands of succOp that are results of currOp
      if (dstOpOperand.get().getDefiningOp() == currOp) {
         // do stuff with dstOpOperand
    }
  }
}
```


I wanted to share a concern regarding memory safety. When the IR is modified, e. g., once the speculator is placed, what happens with the `OpOperand *` that pointed to the operand edge where we now have the speculator? Is that part of memory still pointing to the same `OpOperand`? 

But I did some tests and didn't get any segmentation faults, so it is likely to be memory safe.

lucas-rami commented:
Thanks @asleix for addressing everything quickly :)

(tip: For minor PR comments, you can just do "Resolve conversation" instead of reacting with a +1 to my message. I trust that in those cases you have addressed the requested change and I will not check. Only leave the conversation open if you want me to look at something or want to argue/discuss the change or simply replied to my comment. This allows me to go a bit faster over your updates ;))

> Some of the syntax may have become a bit more verbose. To traverse the IR in a DFS manner, now I use...

You can be shorter and more efficient with
```cpp
for (OpResult res : currOp->getResults()) {
  for (OpOperand &dstOpOperand : res.getUses()) {
    // do stuff with dstOpOperand
  }
}
```

> I wanted to share a concern regarding memory safety. When the IR is modified, e. g., once the speculator is placed, what happens with the OpOperand * that pointed to the operand edge where we now have the speculator? Is that part of memory still pointing to the same OpOperand?

That's a good question, and a scenario where I'm not completely sure of what happens under the hood. I think it is safe to assume that, as long as the operands of an operation do not change, then they will maintain their location in memory (i.e., your `OpOperand*`'s will remain valid). However the story is probably different as soon as you modify them. I would not assume that pointers remain valid after modifications to the operands of an operation i.e., if `p` of type `OpOperand*` holds a pointer to operand `i` of some operation `op`, then after you replace the `i`'th operand of `op` with the result of a new operation, `p` should be assumed to be garbage and never used again. As long as you don't modify any given operand more than once this means you should probably be fine (unless you are for some reason re-using the pointer to some previously replaced operand at any point).

> Finally, regarding moving the CFG code to CFG.h/cpp, my suggestion is to create a PR moving the code after merging this PR. :)

Perfect :)

I let you address the DFG traversal thing and then we can merge!

asleix commented:
Thanks @lucas-rami! 
I have improved the syntax for the DFS traversal with your suggestion. And thanks a lot for the thoughts and insights regarding the OpOperand pointers. Also, in the future, I'll mark as resolved directly for minor comments.

I have merged with main and run the tests to verify correct behavior.
I'll soon open PRs, one for moving code to `CFG.h` and the other for some additions to `DOTPrinter.h` to support speculation.
